### PR TITLE
[VideoPlayer] OpenDefaultStreams: dont close non-CC subtitle streams

### LINKED ISSUE #20404: [Kodi 20][InputStream][Demuxer] DEMUX_SPECIALID_STREAMCHANGE reinitialize and disable the subtitles
### Describe the bug
Here is a clear and concise description of what the problem is:

I am working to the new webvtt subtitles handler, then i have done several tests in various ways,
and i noticed a problem that cause subtitles to be disabled without apparently reasons when you play a video with ISAdaptive add-on.

I have not a full picture in how all demuxer thing works, but there is something wrong somewhere, i try explain how i can

The InputStreamAdaptive add-on (on Nexus branch), has now the new feature to switch streams based on available bandwidth, then in theory should send this DEMUX_SPECIALID_STREAMCHANGE empty packet to kodi demuxer every time that a stream change.

As i have read here
https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-dev-kit/include/kodi/addon-instance/Inputstream.h#L1522-L1534
explains that this reinitialize kodi player every time, then kodi request all streams again

The problem happen that, if you are playing a video with subtitles enabled and ISA do the DEMUX_SPECIALID_STREAMCHANGE callback, kodi disable the current subtitles displayed, then every time you have to open the GUI menu and re-enable the subtitles, again, again and again...infinitely...

OFC this, cause the resettings also of the current subtitle overlay instance loaded (from the log you can see `CDVDSubtitlesLibass` reloaded and the subs data is reparsed every time), this means that every time that a stream change, the full subtitle data will be reparsed in full every time, i find this behaviour excessive...

Maybe a possible solution could be to create a new different "DEMUX_SPECIALID_STREAMCHANGE" that allow resetting only audio/video streams and do not touch the subtitles streams, in order to keep the currently used subtitle overlay instance without recreate it from scratch and then avoid reparse whole subtitle data every time this callback happen

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

### To Reproduce

Hard to reproduce in automatically way depends on bandwitch, type of stream used, server provider speed,
in my case this happen every time with Neftx non-DRM streams so with trailers videos,
then in the log you can see `DEMUX_SPECIALID_STREAMCHANGE` text output

But i think that another similar way could be set in the ISAdaptive settings `manual video stream selection`
then when you play a non-DRM video, you can switch manually via GUI the video streams, in this way you can get the DEMUX_SPECIALID_STREAMCHANGE problem

I have always said of non-DRM videos because i cannot test "manual video stream switch" with Netfx DRM-videos, because currently ISA has something broken (at least on Windows build) and if i try switch the video image will be permanently freezed.

Since this problem can also affect the selection of subtitles language another way to reproduce the problem is with a Dash stream by using regular Kodi20 build/ISA build:
- Play this Dash STRM: [_Test_WebVTT_segmented_HLS.strm.txt](https://github.com/xbmc/xbmc/files/7611699/_Test_WebVTT_segmented_HLS.strm.txt)
- Load an external subtitle file and enable subtitles via GUI
- Slow down your network to make ISA changing the video stream quality
- When ISA will change quality `AddOnLog: inputstream.adaptive: DEMUX_SPECIALID_STREAMCHANGE` will be printed on the Kodi log (keep track in realtime) after that you can see also `Created subtitles overlay codec: ***` that means that has been resetted the subtitle parser instance and requested again subtitle data and re-parsed the subtitle data
- Now (always while in playback) open GUI subtitles settings, you'll see that the external subtitles track is no longer selected and it has re-selected Kodi default subtitle language track
None of this should have happened while changing A/V quality stream

### Debuglog

The debuglog can be found here:
[kodi_specialix.log.txt](https://github.com/xbmc/xbmc/files/7418724/kodi_specialix.log.txt)

### Screenshots 
Here are some links or screenshots to help explain the problem:

## Additional context or screenshots (if appropriate)
Here is some additional context or explanation that might help:

In the InputStream Adaptive the DEMUX_SPECIALID_STREAMCHANGE callback happen here:
https://github.com/xbmc/inputstream.adaptive/blob/Nexus/src/main.cpp#L3846-L3852

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [ ] OSX
 - [x] Windows
 - [ ] Windows UWP

 - Operating system version/name: Windows 10
 - Kodi version: master self-builded

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

### PR DESCRIPTION
**Pull request by CastagnaIT**
## Description

This change allow us when we send `DEMUX_SPECIALID_STREAMCHANGE` to keep the same subtitle opened stream, and avoid that subtitle parser/renderer (libass) instance to be resetted

That otherwise the reinitialization of subtitle parser and the renderer (Libass) will cause side effects like:
- Stored parsed data in libass will be lost, at every reinit will force the re-download(from ISAdaptive)/re-init libass renderer/re-parsing full file while in playback
- In case of too many fonts to load for subtitles will cause a delay in the display of subtitles, therefore nothing will be displayed during (re-)loading
- In case of big subtitle file the parsing could take time on slower cpu, therefore nothing will be displayed during (re-)parsing

## Motivation and context

When we play a DASH/HLS adaptive video streams, InputStream Adaptive send to kodi `DEMUX_SPECIALID_STREAMCHANGE` packet to let know the VP of changed audio/video quality during the playback (some factors determine the change in video quality such as bandwidth), which lead to a forced processing of each component with all side effets of the case

Fix #20404 related to https://github.com/xbmc/inputstream.adaptive/issues/836#issuecomment-977739259
Superseeds #20579 where there is all the discussions/info on this problem

## How has this been tested?

tested with PVR/channel switching to check regressions, comment: https://github.com/xbmc/xbmc/pull/20579#issuecomment-1105094583

tested with ISAdaptive with a dash stream with Webvtt like "file" stream, by forcing adaptive stream change

## What is the effect on users?

Have subtitle working when playing DASH/HLS with adaptive video stream technology

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by CastagnaIT** [comment]
fernet assume that the solution is elsewere, i can say that he is right and i have think a lot for a better solution,
but this PR is actually what at the moment we can do since Kodi now has a subtitle renderer (libass) which in the past was not designed for this use case, a big rework of a lot of things is needed to a best solution

the solution that i was think is:
- rework kodi core in order to start and preserve a single subtitle libass rendering instance from kodi start, until his end
- rework all kodi subtitle parsers to decuple libass rendering from the parser itself, then in theory we could be able to make work the parser with only `ASS_Track` and bound it to the existing rendering instance when needed, this we prevent to reinitialize the libass renderer each time and avoid to reload each time all fonts (cause of delay)
- the InputStream API interface must provide the access for the subtitle parsers, in this way we can parse from InputStreamAdaptive the subtitles provided as file mode (that are no packets) and feed the kodi demuxer as usual for each PTS

this will solve in full the problem but it is a very huge work concerning many things
will probably be done something like this in the future

**Comment by glennguy** [comment]
@CastagnaIT I think there may be a simpler more concise solution in following the suggestion in https://github.com/xbmc/xbmc/pull/20579#issuecomment-1079642663 : (emphasis mine)
> No, this is not the case, at least not in general. This is what I meant with my comment. Subs should be handled like a/v streams. There is one exception with closed captions because those are muxed into the video stream. I guess this is what this old change tried to fix: https://github.com/xbmc/xbmc/pull/6105
***Maybe close subs only in case of cc.***

The change referred to in #6105 is this line addition btw: https://github.com/xbmc/xbmc/blob/2b6cc755abf04aa51192de9f1bfa00dd9ceb075d/xbmc/cores/VideoPlayer/VideoPlayer.cpp#L936

So rather than always closing the sub stream to deal with the possibility of conflicts when opening an external sub stream while CC subs are on, we check if CC subs are currently opened and only act if they are:
```
if (m_CurrentSubtitle.hint.codec == AV_CODEC_ID_EIA_608)
  CloseStream(m_CurrentSubtitle, false);
```

The other code changes are then unnecessary, this becomes a one liner and PR topic becomes "OpenDefaultStreams - Don't close non-CC subtitle streams"

This conforms to the stated design requirement in https://github.com/xbmc/xbmc/pull/20579#issuecomment-1079657035 : 
> The design is as follows: VP can call OpenDefaultStreams as many times as it wants to, for what reason ever. This must not do any harm to playing streams if nothing has changed.

**Comment by CastagnaIT** [comment]
i have tested your proposal and i can confirm that works
so ISA not re-download subtitles and subtitle codec/libass renderer is kept

in case `PredicateSubtitlePriority` will do side effects for some use cases will be fixed on other times
the change is therefore even smaller than before

**Comment by CastagnaIT** [comment]
updated title and description in accordance with changes

**Review by glennguy — APPROVED** [review]

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 1]
```diff
@@ -933,7 +933,10 @@ void CVideoPlayer::OpenDefaultStreams(bool reset)
                                 m_processInfo->GetVideoSettings().m_SubtitleStream,
                                 m_processInfo->GetVideoSettings().m_SubtitleOn);
   valid = false;
-  CloseStream(m_CurrentSubtitle, false);
+  // We need to close CC subtitles to avoid conflicts with external sub stream
+  if (m_CurrentSubtitle.hint.codec == AV_CODEC_ID_EIA_608)
```
Comment (enen92) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:937]: Is there other way to refer to cc subtitles generically used in the codebase? What about for example EIA_708?

Comment (CastagnaIT) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp:937]: on our libavcodec\codec_id.h seems there is only that one
however also in CC demuxer is specified:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxCC.cpp#L320

Comment (enen92) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp:937]: Yeah we seem to reuse the same codec id for both formats

**Review by enen92 — APPROVED** [review]

**Comment by enen92** [comment]
@FernetMenta can we have your okay here if you have a bit of time? Simple one liner (same context of other attempts)

**Review by phunkyfish — APPROVED** [review]

**Comment by FernetMenta** [comment]
The change that introduced AV_CODEC_ID_EIA_608 for CDVDDemuxCC was already wrong. Hence comparing for this codec ID here is wrong too. If you go down this road further, you will run into more problems sooner or later. 

In this case here you should test for STREAM_SOURCE_VIDEOMUX.

**Comment by CastagnaIT** [comment]
changed condition to compare source->STREAM_SOURCE_VIDEOMUX

**Comment by CastagnaIT** [comment]
jenkins build this please

**Pull request merged**